File: Your-Space-Backend/YourSpace.Services/Helper/RelationInverseResolver.cs
using YourSpace.Data.Enums;

namespace YourSpace.Services.Helper;

// Stateless domain lookup (not a DI service, same placement precedent as LocalizedTextResolver) —
// resolves the inverse of a PersonRelationship row. `subjectGender` is always the gender of the
// row's own PersonId (the "Y" in "Y's father is X"), never RelatedPersonId, regardless of
// direction — see relationship-engine-2026-08-05.md's inverse-derivation table.
//
// Known limitation, not silently handled: Nephew/Niece -> Uncle/Aunt loses the Maternal/Paternal
// distinction (can't tell which side without walking the rest of the graph). Defaults to Paternal
// on that one ambiguous direction — the far more common flow (Uncle/Aunt -> Nephew/Niece) is
// unaffected and always correct.
public static class RelationInverseResolver
{
    public static RelationType Resolve(RelationType original, Gender subjectGender) =>
        subjectGender == Gender.Male ? MaleInverse[original] : FemaleInverse[original];

    private static readonly Dictionary<RelationType, RelationType> MaleInverse = new()
    {
        [RelationType.Father] = RelationType.Son,
        [RelationType.Mother] = RelationType.Son,
        [RelationType.Son] = RelationType.Father,
        [RelationType.Daughter] = RelationType.Father,
        [RelationType.Brother] = RelationType.Brother,
        [RelationType.Sister] = RelationType.Brother,
        [RelationType.Husband] = RelationType.Wife,
        [RelationType.Wife] = RelationType.Husband,
        [RelationType.UncleMaternal] = RelationType.Nephew,
        [RelationType.AuntMaternal] = RelationType.Nephew,
        [RelationType.UnclePaternal] = RelationType.Nephew,
        [RelationType.AuntPaternal] = RelationType.Nephew,
        [RelationType.Nephew] = RelationType.UnclePaternal,
        [RelationType.Niece] = RelationType.UnclePaternal
    };

    private static readonly Dictionary<RelationType, RelationType> FemaleInverse = new()
    {
        [RelationType.Father] = RelationType.Daughter,
        [RelationType.Mother] = RelationType.Daughter,
        [RelationType.Son] = RelationType.Mother,
        [RelationType.Daughter] = RelationType.Mother,
        [RelationType.Brother] = RelationType.Sister,
        [RelationType.Sister] = RelationType.Sister,
        [RelationType.Husband] = RelationType.Wife,
        [RelationType.Wife] = RelationType.Husband,
        [RelationType.UncleMaternal] = RelationType.Niece,
        [RelationType.AuntMaternal] = RelationType.Niece,
        [RelationType.UnclePaternal] = RelationType.Niece,
        [RelationType.AuntPaternal] = RelationType.Niece,
        [RelationType.Nephew] = RelationType.AuntPaternal,
        [RelationType.Niece] = RelationType.AuntPaternal
    };
}

File: Your-Space-Mobile/lib/features/people/data/relation_inverse_resolver.dart
import 'package:your_space_mobile/core/entities/gender.dart';
import 'package:your_space_mobile/core/entities/relation_type.dart';

/// Mobile-side port of the backend's `RelationInverseResolver` — needed
/// because Tier 2's optimistic local write (row 9.13) inserts the
/// auto-derived inverse row *before* the server confirms it, so the local
/// drift cache needs the same deterministic mapping the backend uses to
/// avoid ever showing an inaccurate placeholder type between the write and
/// the sync. `subjectGender` is always the gender of the row's own
/// `personId` (the "Y" in "Y's father is X"), never `relatedPersonId`,
/// regardless of direction.
///
/// Known limitation, inherited from the backend: Nephew/Niece -> Uncle/Aunt
/// loses the Maternal/Paternal distinction (can't tell which side without
/// walking the rest of the graph) — defaults to Paternal, same as the
/// backend.
class RelationInverseResolver {
  const RelationInverseResolver._();

  static RelationType resolve(RelationType original, Gender subjectGender) =>
      subjectGender == Gender.male ? _maleInverse[original]! : _femaleInverse[original]!;

  static const _maleInverse = {
    RelationType.father: RelationType.son,
    RelationType.mother: RelationType.son,
    RelationType.son: RelationType.father,
    RelationType.daughter: RelationType.father,
    RelationType.brother: RelationType.brother,
    RelationType.sister: RelationType.brother,
    RelationType.husband: RelationType.wife,
    RelationType.wife: RelationType.husband,
    RelationType.uncleMaternal: RelationType.nephew,
    RelationType.auntMaternal: RelationType.nephew,
    RelationType.unclePaternal: RelationType.nephew,
    RelationType.auntPaternal: RelationType.nephew,
    RelationType.nephew: RelationType.unclePaternal,
    RelationType.niece: RelationType.unclePaternal,
  };

  static const _femaleInverse = {
    RelationType.father: RelationType.daughter,
    RelationType.mother: RelationType.daughter,
    RelationType.son: RelationType.mother,
    RelationType.daughter: RelationType.mother,
    RelationType.brother: RelationType.sister,
    RelationType.sister: RelationType.sister,
    RelationType.husband: RelationType.wife,
    RelationType.wife: RelationType.husband,
    RelationType.uncleMaternal: RelationType.niece,
    RelationType.auntMaternal: RelationType.niece,
    RelationType.unclePaternal: RelationType.niece,
    RelationType.auntPaternal: RelationType.niece,
    RelationType.nephew: RelationType.auntPaternal,
    RelationType.niece: RelationType.auntPaternal,
  };
}
